File: 0054-spiral-matrix/0054-spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        // find out  Row size and colm size 
        int row = matrix.size();
        int col = matrix[0].size();
        vector<int> temp;
        // Intialise all the Index value as you are printing
        int StartRow = 0;
        int Endcolm = col - 1;
        int EndRow = row - 1;
        int StartCol = 0;
        int total  = row * col;
        int count = 0;
        while(count < total)
        {
            for(int j = StartCol ; j <= Endcolm && count < total;j++)
            {
                temp.push_back(matrix[StartRow][j]);
                count++;
            }
            StartRow++;
            for(int j = StartRow; j <= EndRow  && count < total; j++)
            {
                temp.push_back(matrix[j][Endcolm]);
                count++;
            }
            Endcolm--;
            for(int j = Endcolm; j >= StartCol && count < total;  j--)
            {
                temp.push_back(matrix[EndRow][j]);
                count++;
            }
            EndRow--;
            for(int j = EndRow; j >= StartRow && count < total; j--)
            {
                temp.push_back(matrix[j][StartCol]);
                count++;
            }
            StartCol++;
            
        }
        return temp;
    }
};

File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """

        res = []
        row = len(matrix)
        col = len(matrix[0])

        left = 0
        right = col - 1
        top = 0
        bottom = row - 1

        while left <= right and top <= bottom:

            # first loop..
            for j in range(left , right + 1 ):

                res.append(matrix[top][j])
            top += 1
            # second loop..
            for j in range(top , bottom + 1):

                res.append(matrix[j][right])
            right -= 1

            # 3rd loop..

            if top <= bottom:

                for j in range(right , left - 1 , -1):

                    res.append(matrix[bottom][j])

                bottom -= 1

            # 4th loop..
            if left <= right:

                for j in range(bottom, top-1 , -1):
                    res.append(matrix[j][left])

                left += 1

        return res









       